File: findMinHeightTrees.py
from typing import List


# 最小高度树
class Solution:
    def findMinHeightTrees(self, n: int, edges: List[List[int]]) -> List[int]:
        adj = [set() for i in range(n)]

        for e in edges:
            adj[e[0]].add(e[1])
            adj[e[1]].add(e[0])
        queue = []

        for i in range(n):
            if len(adj[i]) <= 1:
                queue.append(i)
        pre_queue = queue

        while len(queue) > 0:
            pre_queue = queue
            queue = []
            queue

            for node in pre_queue:
                for np in adj[node]:
                    adj[np].remove(node)

                    if len(adj[np]) == 1:
                        queue.append(np)

        return pre_queue


n = 4
edges = [[1, 0], [1, 2], [1, 3]]
# n = 6
# edges = [[3,0],[3,1],[3,2],[3,4],[5,4]]
n = 1
edges = []
sl = Solution()
print(sl.findMinHeightTrees(n, edges))

File: findMinHeightTrees.java
import java.util.HashSet;
import java.util.Set;
import java.util.Arrays;
import java.util.Stack;
import java.util.Queue;
import java.util.LinkedList;
import java.util.List;
import java.util.ArrayList;

class Solution {

    public List<Integer> findMinHeightTrees(int n, int[][] edges) {
        List<Set<Integer>> adjList=new ArrayList<>();
        for(int i=0;i<n;i++) {
            adjList.add(new HashSet<>());
        }
        for (int[] edge : edges) {
            adjList.get(edge[0]).add(edge[1]);
            adjList.get(edge[1]).add(edge[0]);
        }
        List<Integer> queue=new ArrayList<>();
        for(int i=0;i<n;i++) {
            if(adjList.get(i).size()<=1) {
                queue.add(i);
            }
        }
        List<Integer> preQueue=queue;
        while(queue.size()>0) {
            preQueue=queue;
            queue=new ArrayList<>();
            for(int node:preQueue) {
                for(int np:adjList.get(node)) {
                    adjList.get(np).remove(node);
                    if(adjList.get(np).size()==1) {
                        queue.add(np);
                    }
                }
            }
        }
        return preQueue;
    }

    public static void main(String[] args) {
        int n = 4;
        // int[][] edges = {{1,0},{1,2},{1,3}};
        n = 6;
        int[][] edges = {{3,0},{3,1},{3,2},{3,4},{5,4}};
        Solution sl=new Solution();
        System.out.println(sl.findMinHeightTrees(n,edges));
    }
}

